File: platform/src/java/com/xcase/rest/generator/swagger/SwaggerProxyGenerator.java
package com.xcase.rest.generator.swagger;

import com.xcase.rest.generator.RESTServiceDefinition;
import com.xcase.rest.generator.RESTProxyGenerator;
import com.xcase.rest.generator.IAPIProxySettingsEndpoint;
import com.xcase.common.impl.simple.core.CommonHTTPManager;
import com.xcase.rest.generator.ClassDefinition;
import com.xcase.rest.generator.EnumDefinition;
import com.xcase.rest.generator.Operation;
import com.xcase.rest.generator.Parameter;
import com.xcase.rest.generator.ParameterIn;
import com.xcase.rest.generator.ProxyDefinition;
import com.xcase.rest.generator.RESTParser;
import com.xcase.rest.generator.TypeDefinition;
import java.io.File;
import java.io.PrintWriter;
import java.lang.invoke.MethodHandles;
import java.util.*;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public abstract class SwaggerProxyGenerator extends RESTProxyGenerator {
    /**
     * log4j object.
     */
    protected static final Logger LOGGER = LogManager.getLogger(MethodHandles.lookup().lookupClass());

    public abstract RESTServiceDefinition generateSourceString(IAPIProxySettingsEndpoint[] endpoints) throws Exception;
    public abstract RESTServiceDefinition generateSourceString(IAPIProxySettingsEndpoint endpoint, String document, String username, String password, String tenant) throws Exception;
    public abstract RESTServiceDefinition generateSourceString(IAPIProxySettingsEndpoint endpoint, String document) throws Exception;
    public abstract RESTServiceDefinition generateSourceString(String document) throws Exception;
    public abstract List<RESTServiceDefinition> generateSourceStringForRestServiceDefinitionList(IAPIProxySettingsEndpoint swaggerApiProxySettingsEndPoint, String swaggerDocument, String username, String password, String tenant) throws Exception;
}

File: XCase.Swagger.ProxyGenerator/Generator/SwaggerProxyGenerator.cs
namespace XCase.REST.ProxyGenerator
{
    using System;
    using System.Collections.Concurrent;
    using System.Collections.Generic;
    using System.IO;
    using System.Linq;
    using System.Net;
    using System.Net.Http;
    using System.Reflection;
    using System.Security;
    using System.Text;
    using System.Threading.Tasks;
    using Microsoft.CSharp;
    using Microsoft.Owin.Testing;
    using log4net;
    using Newtonsoft.Json;
    using XCase.ProxyGenerator;
    using XCase.REST.ProxyGenerator.Generator;
    using XCase.REST.ProxyGenerator.Swagger;

    public abstract class SwaggerProxyGenerator : RESTProxyGenerator
    {
        #region Logger Setup

        /// <summary>
        /// A log4net log instance.
        /// </summary>
        private static readonly ILog Log = log4net.LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);

        #endregion

        public static StringBuilder SourceStringBuilder { get; set; }
        public static ConcurrentDictionary<IAPIProxySettingsEndpoint, string> swaggerDocDictionary = new ConcurrentDictionary<IAPIProxySettingsEndpoint, string>();

        public override abstract IServiceDefinition GenerateSourceString(IAPIProxySettingsEndpoint[] endpoints);
        public override abstract IServiceDefinition GenerateSourceString(IAPIProxySettingsEndpoint endpoint, string document, string username, string password, string tenant);
        public override abstract IServiceDefinition GenerateSourceString(IAPIProxySettingsEndpoint endpoint, string document);
        public override abstract IServiceDefinition GenerateSourceString(string document);

        public static async Task GetEndpointSwaggerDoc(string requestUri, IAPIProxySettingsEndpoint endPoint)
        {
            Log.Debug("starting GetEndpointSwaggerDoc()");
            string swaggerString = null;
            System.Net.WebRequest webRequest = System.Net.WebRequest.Create(requestUri);
            Log.Debug("created webRequest");
            using (WebResponse webResponse = await webRequest.GetResponseAsync().ConfigureAwait(false))
            {
                Log.Debug("got webResponse");
                Stream webResponseStream = webResponse.GetResponseStream();
                StreamReader webResponseStreamReader = new StreamReader(webResponseStream);
                swaggerString = await webResponseStreamReader.ReadToEndAsync().ConfigureAwait(false);
            }

            if (swaggerString == null)
            {
                throw new Exception(string.Format("Error downloading from: {0}", endPoint.GetUrl()));
            }

            Log.DebugFormat("downloaded: {0}", requestUri);
            swaggerDocDictionary.GetOrAdd(endPoint, swaggerString);
            Log.Debug("finishing GetEndpointSwaggerDoc()");
        }
    }
}
